functionalDB/chapter.md

```diff
@@ -0,0 +1,1138 @@
+         cleaned-result-clauses (map (partial mask-path-leaf-with-items relevant-items)
+                                     result-clauses)] 
+     (filter #(not-empty (last %)) cleaned-result-clauses)))
+````
+This function starts by applying the predicate clauses on the previously-chosen index. Each application of a predicate clause on an index returns a _result clause_. 
+
+The main characteristics of a result are:
+
+1. It is built of three items, each from a different level of the index, and each passed its respective predicate. 
+2. The order of items matches the index's levels structure (predicate clause are always in EAV order). The re-ordering is done when applying the index's *from-eav* on the predicate clause. 
+3. The metadata of the predicate clause is attached to it. 
+
+All of this is done in the function *filter-index*
+
+````clojure
+(defn filter-index [index predicate-clauses]
+   (for [pred-clause predicate-clauses
+         :let [[lvl1-prd lvl2-prd lvl3-prd] (apply (from-eav index) pred-clause)] 
+         [k1 l2map] index  ; keys and values of the first level
+         :when (try (lvl1-prd k1) (catch Exception e false))  
+         [k2  l3-set] l2map  ; keys and values of the second level
+         :when (try (lvl2-prd k2) (catch Exception e false))
+         :let [res (set (filter lvl3-prd l3-set))] ]
+     (with-meta [k1 k2 res] (meta pred-clause))))
+````
+Assuming the query was executed on July 4th, the results of executing it on the above data are:
+<table>
+<tr>
+<td>Result Clause</td><td>Result Meta</td>
+</tr>
+<tr>
+<td>[:likes Pizza #{1}]</td><td>["?e" nil nil]</td>
+</tr>
+<tr>
+<td>[:name USA #{1}]</td><td>["?e" nil "?nm"]</td>
+</tr>
+<tr>
+<td>[:speak "English" #{1, 3}]</td><td>["?e" nil nil]</td>
+</tr>
+<tr>
+<td>[:birthday "July 4, 1776" #{1}]</td><td>["?e" nil "?bd"]</td>
+</tr>
+<tr>
+<td>[:name France #{2}]</td><td>["?e" nil "?nm"]</td>
+</tr>
+<tr>
+<td>[:birthday "July 14, 1789" #{2}]</td><td>["?e" nil "?bd"]</td>
+</tr>
+<tr>
+<td>[:name Canada #{3}]</td><td>["?e" nil "?nm"]</td>
+</tr>
+<tr>
+<td>[:birthday "July 1, 1867" {3}]</td><td>["?e" nil "?bd"]</td>
+</tr>
+</table>
+Once we have produced all of the result clauses, we need to perform an *AND* operation between them. This is done by finding all of the elements that passed all the predicate clauses:
+
+````clojure
+(defn items-that-answer-all-conditions [items-seq num-of-conditions]
+   (->> items-seq ; take the items-seq
+         (map vec) ; make each collection (actually a set) into a vector
+         (reduce into []) ;reduce all the vectors into one vector
+         (frequencies) ; count for each item in how many collections (sets) it was in
+         (filter #(<= num-of-conditions (last %))) ;items that answered all conditions
+         (map first) ; take from the duos the items themselves
+         (set))) ; return it as set
+````
+
+In our example, the result of this step is a set that holds the value *1*. 

```

Comment:
Maybe clarify why it holds the value 1 (although I do understand why, so maybe that's not necessary)